util/env_posix.cc

```diff
@@ -305,15 +306,19 @@ class PosixEnv : public Env {
     }
     if (options.use_mmap_writes && !forceMmapOff_) {
       result->reset(new PosixMmapFile(fname, fd, page_size_, options));
-    } else if (options.use_direct_writes && !options.use_mmap_writes) {
+    } else if (use_direct) {
 #ifdef OS_MACOSX
       if (fcntl(fd, F_NOCACHE, 1) == -1) {
         close(fd);
         s = IOError(fname, errno);
         return s;
       }
 #endif
-      result->reset(new PosixWritableFile(fname, fd, options));
+      if (fname.rfind(".log") != std::string::npos) {
+        result->reset(new WALWritableFile(fname, fd, options));
+      } else {
```

Comment:
that was the first thing i tried, and in fact is what im running in production.  however, apart from some minor differences like the way Allocate() needs to work due to fallocate() problems in XFS, the biggest reason was because the design of the current one does the pwrite() in PositionedAppend() which fundamentally means: (1) you'll never get a write larger than 32k, because thats the rocksdb block size, and (2) you're guaranteed to have a minimum of 2 writes on every commit (header + data), and (3) you're now essentially ignoring the WriteOption.sync flag, since its going to disk anyway given that its O_DIRECT.  with the changes im proposing, the journaling speed goes up by a considerable amount in my workload, which is about 2/3 sync commits and 1/3 allowed to be lossy.